### PR TITLE
Split Less and SCSS parsing into different parsers

### PR DESCRIPTION
**Pull request by lydell**
Now, .less files are always parsed with postcss-less, and .scss files
areare always parsed with postcss-scss. This:

- Is less hacky.
- Is meant to avoid issues like #2829.
- Is probably more performant.

`parser: "postcss"` and `--parser postcss` continue to work like before:
First trying postcss-less, and if that fails, postcss-scss, unless a
regex says that we should try in the opposite order. The new values for
the parser option are "postcss-less" and "postcss-scss".

### CODE REVIEW

**Comment by azz** [comment]
I feel like the `parser` option is a bit of a leaky abstraction. It would be nice if the option was called `language` or something like that so you can pass `--language scss` and we'd figure out which parser to use.

`parser` | `language` 
-|-
babylon|javascript
flow|flow
typescript (actually `typescript-eslint-parser`)|typescript
graphql|graphql
postcss|css
postcss-scss|scss
postcss-less|less
parse5|html
json (actually `babylon.parseExpression`)|json


Unfortunately changing this would obviously be an API breaking change.

**Comment by lydell** [comment]
I agree that the parser option is strange and could be better. But I don't think it should affect this PR in any way.

Btw, the postcss parts of your table should look more like this:

`parser` | `language` 
-|-
postcss (actually postcss-less+postcss-scss+postcss-media-query-parser+postcss-selector-parser+postcss-values-parser)|css
postcss-scss (actually like css but minus postcss-less)|scss
postcss-less (actually like css but minus postcss-scss)|less

**Comment by vjeux** [comment]
Agreed, parser was designed before we intended to support more than just pure JavaScript.

I think that it makes sense to introduce a language option and we figure out a way to make it backwards compatible.

File: website/pages/playground/index.html [hunk 1]
```diff
@@ -359,7 +359,7 @@
         </div>
         <div class="options last">
           <label>--trailing-comma <select id="trailingComma"><option value="none">none</option><option value="es5">es5</option><option value="all">all</option></select></label>
-          <label>--parser <select id="parser"><option value="babylon">babylon</option><option value="flow">flow</option><option value="typescript">typescript</option><option value="postcss">postcss</option><option value="json">json</option><option value="graphql">graphql</option></select></label>
+          <label>--parser <select id="parser"><option value="babylon">babylon</option><option value="flow">flow</option><option value="typescript">typescript</option><option value="postcss">postcss</option><!-- TODO: Enable these when prettier@>1.7.0 has been released. <option value="postcss-less">postcss-less</option><option value="postcss-scss">postcss-scss</option>--><option value="json">json</option><option value="graphql">graphql</option></select></label>
```
Comment (vjeux) [inline website/pages/playground/index.html]: 1.8.0 now

Comment (azz) [reply website/pages/playground/index.html]: Exclusive `>1.7.0` is technically correct :wink:

File: tests_integration/__tests__/__snapshots__/early-exit.js.snap [hunk 2]
```diff
@@ -84,7 +84,7 @@ Format options:
 
   --no-bracket-spacing     Do not print spaces between brackets.
   --jsx-bracket-same-line  Put > on the last line instead of at a new line.
-  --parser <flow|babylon|typescript|postcss|json|graphql>
+  --parser <flow|babylon|typescript|postcss|postcss-less|postcss-scss|json|graphql>
```
Comment (vjeux) [inline tests_integration/__tests__/__snapshots__/early-exit.js.snap]: Suggestion: let's just name it `less` and `scss`. And allow both `postcss` and `css` to work and only mention `css` in the docs.

We're unlikely going to ever support anything else than postcss to parse our CSS and at the end of the day the users don't care, they just want to parse their css/less/sass.

**Comment by azz** [comment]
> But I don't think it should affect this PR in any way.
Indeed. Opened #2846 for discussion.

File: tests_integration/__tests__/__snapshots__/early-exit.js.snap (hunk 2 above)
Comment (lydell) [reply tests_integration/__tests__/__snapshots__/early-exit.js.snap]: Do you mean that it is unlikely that we'll support two different parsers for the same language, like we do for JS with both Babylon and Flow? Or that it is unlikely that we'll every switch from postcss to something else?

I have been researching switching away from postcss:

- [css-tree](https://github.com/csstree/csstree) for CSS
- [less](https://github.com/less/less.js) or [gonzales-pe](https://github.com/tonyganch/gonzales-pe/) for Less
- [dart-sass](https://github.com/sass/dart-sass/) or [gonzales-pe](https://github.com/tonyganch/gonzales-pe/) for SCSS (see also https://github.com/sass/dart-sass/issues/88)

Comment (azz) [reply tests_integration/__tests__/__snapshots__/early-exit.js.snap]: If we did switch away, we probably wouldn't want to keep supporting postcss (another advantage of using --language)

Comment (vjeux) [reply tests_integration/__tests__/__snapshots__/early-exit.js.snap]: Yeah, what I mean is that we just name it with the language, so we don't get stuck to a specific parser if we want to switch.

**Comment by lydell** [comment]
Updates:

- postcss-less → less
- postcss-scss → scss
- deprecated postcss in favor of css

File: website/static/playground.js [hunk 3]
```diff
@@ -238,6 +238,8 @@ function replaceHash(hash) {
 
 function getCodemirrorMode(options) {
   switch (options.parser) {
+    // TODO: Remove the "postcss" case when prettier@>1.7.0 is released.
```
Comment (suchipi) [inline website/static/playground.js]: Is this a prerelease task, or a postrelease task? (Or doesn't matter)

Comment (lydell) [reply website/static/playground.js]: post

**Comment by ismail-syed** [comment]
@lydell As per your comment: `deprecated postcss in favor of css`
Do the [parser](https://prettier.io/docs/en/options.html#parser) docs need updating?

**Comment by lydell** [comment]
@ismail-syed They do, well spotted!

**Pull request merged**